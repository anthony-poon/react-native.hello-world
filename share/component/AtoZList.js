import {StyleSheet, Text, View, SectionList} from "react-native";
import React from "react";
import PropTypes from "prop-types";
import _ from "lodash";
import {TextStyle, Padding, Margin} from "../../share/style"
import {COLOR_LIGHT} from "../style/var";
import Toast from 'react-native-root-toast';

export default class AtoZList extends React.Component {
    sectionListRef = React.createRef();
    keyListRef = React.createRef();
    listOffsetY = 0;
    keysOffsetY = [];
    throttledTouchHandler = _.throttle(this.handListKeyTouch.bind(this), 50, {
        leading: true,
        trailing: false
    });


    state = {
        currKey: null
    };

    handListKeyTouch(evt) {
        let index = _.findIndex(this.keysOffsetY, offset => {
            return this.listOffsetY + offset > evt.nativeEvent.pageY
        });
        if (index === -1) {
            index = this.keysOffsetY.length - 1
        }
        this.setState({
            currKey: index
        });
        this.sectionListRef.current.scrollToLocation({
            animated: true,
            sectionIndex: index,
            itemIndex: 0,
            viewPosition: 0
        })
    }

    render() {
        let {
            listKeys,
            data,
            renderItem,
            renderSectionHeader,
            onPartition,
            onScrollToIndexFailed,
            ...rest
        } = this.props;
        listKeys = [
            ...listKeys,
            "#"
        ];
        const partition = Array.from(Array(listKeys.length), () => []);
        data.forEach(d => {
            // Use provided partition function to determine which session to put into. if return false or null, put into last
            const key = onPartition(d);
            if (isNaN(key) || key < 0 || key >= listKeys.length) {
                throw new Error("Invalid return value from onPartition function. Must return an integer between 0 and length of list keys - 1 or false")
            }
            if (key !== false && key !== null) {
                partition[key].push(d);
            } else {
                partition[partition.length - 1].push(d)
            }
        });

        const sections = listKeys.map((k, index) => ({
            title: k,
            data: partition[index]
        }));
        return (
            <View style={styles.container}>
                {
                    this.state.currKey !== null && (
                        <Toast
                            position={Toast.positions.CENTER}
                            animation={true}
                            opacity={0.5}
                            visible={true}
                        >
                            {listKeys[this.state.currKey]}
                        </Toast>
                    )
                }

                <SectionList
                    {...rest}
                    ref={this.sectionListRef}
                    style={styles.sectionList}
                    renderItem={renderItem}
                    renderSectionHeader={renderSectionHeader}
                    onScrollToIndexFailed={onScrollToIndexFailed}
                    sections={sections}
                    keyExtractor={(item, index) => item + index}
                />
                <View style={styles.listKeyWrapper}
                      ref={this.keyListRef}
                      onLayout={evt => {
                          this.keyListRef.current.measure((fx, fy, width, height, px, py) => {
                              this.listOffsetY = py;
                          })
                      }}
                      onStartShouldSetResponder={evt =>  true}
                      onMoveShouldSetResponder ={evt =>  true}
                      onResponderGrant={(evt) => {
                          evt.persist();
                          this.throttledTouchHandler(evt);
                      }}
                      onResponderMove={(evt) => {
                          evt.persist();
                          this.throttledTouchHandler(evt);
                      }}
                      onResponderRelease={evt => {
                          this.setState({
                              currKey: null,
                          })
                      }}
                >
                    {
                        listKeys.map((key, index) => (
                            <View key={index}
                                  onLayout={evt => {
                                      this.keysOffsetY.push(evt.nativeEvent.layout.y)
                                  }}
                            >
                                <Text style={styles.listKey}>{key}</Text>
                            </View>
                        ))
                    }
                </View>
            </View>
        );
    }
}

AtoZList.defaultProps = {
    listKeys: _.range(65, 91).map(ascii => String.fromCharCode(ascii)),
    renderItem: ({item, index, section}) => (
        <View style={styles.listItem}>
            <Text key={index}>{item}</Text>
        </View>
    ),
    renderSectionHeader: ({section: {title}}) => (
        <Text style={styles.sectionTitle}>{title}</Text>
    ),
    onScrollToIndexFailed: (error) => {
        console.log("onScrollToIndexFailed called. Ignoring error. override the method to handle the method");
    }
};

AtoZList.propTypes = {
    listKeys: PropTypes.arrayOf(PropTypes.string),
    renderItem: PropTypes.func,
    renderSectionHeader: PropTypes.func,
    data: PropTypes.arrayOf(PropTypes.any).isRequired,
    onPartition: PropTypes.func.isRequired
};

const styles = StyleSheet.create({
    container: {
        flexDirection: "row"
    },
    sectionList: {
        flex: 1,
    },
    sectionTitle: {
        ...Padding.py1,
        ...Margin.mx2,
        ...TextStyle.textPrimary,
        ...TextStyle.sm,
        justifyContent: "flex-start",
        flex: 1,
        borderBottomWidth: 1,
        borderColor: COLOR_LIGHT,

    },
    listItem: {
        ...Padding.py3,
        ...Margin.mx2,
        borderBottomWidth: 1,
        borderColor: COLOR_LIGHT
    },
    listKeyWrapper: {
        ...Padding.py3,
        alignItems: "center",
    },
    listKey: {
        ...TextStyle.sm,
        ...Padding.py1,
        ...Padding.px1,
    }
});