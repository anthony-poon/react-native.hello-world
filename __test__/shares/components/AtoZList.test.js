import React from 'react';
import renderer from 'react-test-renderer'
import {shallow} from "enzyme";
import AtoZList from "../../../views/contact-list/components/AtoZList";
import _ from "lodash";

describe("<AtoZList/>", () => {
    it("rendered correctly", () => {
        const tree = shallow(<AtoZList data={[]} onPartition={() => {}}/>);
        expect(tree).toMatchSnapshot()
    });

    it("rendered correctly with items", () => {
        const input = _.range(65, 91).map(ascii => String.fromCharCode(ascii) + String.fromCharCode(ascii) + String.fromCharCode(ascii));
        const tree = shallow(
            <AtoZList
                data={input}
                onPartition={data => {
                    const ascii = data.toUpperCase().charCodeAt(0);
                    if (ascii >= 65 && ascii < 91) {
                        return ascii - 65;
                    } else {
                        return false;
                    }
                }}/>
        );
        expect(tree).toMatchSnapshot()
    });
});