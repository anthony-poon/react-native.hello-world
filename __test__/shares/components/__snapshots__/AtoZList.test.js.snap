// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AtoZList/> rendered correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <SectionList
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onEndReachedThreshold={2}
    onScrollToIndexFailed={[Function]}
    renderItem={[Function]}
    renderSectionHeader={[Function]}
    scrollEventThrottle={50}
    sections={
      Array [
        Object {
          "data": Array [],
          "title": "A",
        },
        Object {
          "data": Array [],
          "title": "B",
        },
        Object {
          "data": Array [],
          "title": "C",
        },
        Object {
          "data": Array [],
          "title": "D",
        },
        Object {
          "data": Array [],
          "title": "E",
        },
        Object {
          "data": Array [],
          "title": "F",
        },
        Object {
          "data": Array [],
          "title": "G",
        },
        Object {
          "data": Array [],
          "title": "H",
        },
        Object {
          "data": Array [],
          "title": "I",
        },
        Object {
          "data": Array [],
          "title": "J",
        },
        Object {
          "data": Array [],
          "title": "K",
        },
        Object {
          "data": Array [],
          "title": "L",
        },
        Object {
          "data": Array [],
          "title": "M",
        },
        Object {
          "data": Array [],
          "title": "N",
        },
        Object {
          "data": Array [],
          "title": "O",
        },
        Object {
          "data": Array [],
          "title": "P",
        },
        Object {
          "data": Array [],
          "title": "Q",
        },
        Object {
          "data": Array [],
          "title": "R",
        },
        Object {
          "data": Array [],
          "title": "S",
        },
        Object {
          "data": Array [],
          "title": "T",
        },
        Object {
          "data": Array [],
          "title": "U",
        },
        Object {
          "data": Array [],
          "title": "V",
        },
        Object {
          "data": Array [],
          "title": "W",
        },
        Object {
          "data": Array [],
          "title": "X",
        },
        Object {
          "data": Array [],
          "title": "Y",
        },
        Object {
          "data": Array [],
          "title": "Z",
        },
        Object {
          "data": Array [],
          "title": "#",
        },
      ]
    }
    stickySectionHeadersEnabled={true}
    style={
      Object {
        "flex": 1,
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
  <View
    onLayout={[Function]}
    onMoveShouldSetResponder={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "paddingBottom": 16,
        "paddingTop": 16,
      }
    }
  >
    <View
      key="0"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        A
      </Text>
    </View>
    <View
      key="1"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        B
      </Text>
    </View>
    <View
      key="2"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        C
      </Text>
    </View>
    <View
      key="3"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        D
      </Text>
    </View>
    <View
      key="4"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        E
      </Text>
    </View>
    <View
      key="5"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        F
      </Text>
    </View>
    <View
      key="6"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        G
      </Text>
    </View>
    <View
      key="7"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        H
      </Text>
    </View>
    <View
      key="8"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        I
      </Text>
    </View>
    <View
      key="9"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        J
      </Text>
    </View>
    <View
      key="10"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        K
      </Text>
    </View>
    <View
      key="11"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        L
      </Text>
    </View>
    <View
      key="12"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        M
      </Text>
    </View>
    <View
      key="13"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        N
      </Text>
    </View>
    <View
      key="14"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        O
      </Text>
    </View>
    <View
      key="15"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        P
      </Text>
    </View>
    <View
      key="16"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        Q
      </Text>
    </View>
    <View
      key="17"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        R
      </Text>
    </View>
    <View
      key="18"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        S
      </Text>
    </View>
    <View
      key="19"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        T
      </Text>
    </View>
    <View
      key="20"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        U
      </Text>
    </View>
    <View
      key="21"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        V
      </Text>
    </View>
    <View
      key="22"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        W
      </Text>
    </View>
    <View
      key="23"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        X
      </Text>
    </View>
    <View
      key="24"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        Y
      </Text>
    </View>
    <View
      key="25"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        Z
      </Text>
    </View>
    <View
      key="26"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        #
      </Text>
    </View>
  </View>
</View>
`;

exports[`<AtoZList/> rendered correctly with items 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <SectionList
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onEndReachedThreshold={2}
    onScrollToIndexFailed={[Function]}
    renderItem={[Function]}
    renderSectionHeader={[Function]}
    scrollEventThrottle={50}
    sections={
      Array [
        Object {
          "data": Array [
            "AAA",
          ],
          "title": "A",
        },
        Object {
          "data": Array [
            "BBB",
          ],
          "title": "B",
        },
        Object {
          "data": Array [
            "CCC",
          ],
          "title": "C",
        },
        Object {
          "data": Array [
            "DDD",
          ],
          "title": "D",
        },
        Object {
          "data": Array [
            "EEE",
          ],
          "title": "E",
        },
        Object {
          "data": Array [
            "FFF",
          ],
          "title": "F",
        },
        Object {
          "data": Array [
            "GGG",
          ],
          "title": "G",
        },
        Object {
          "data": Array [
            "HHH",
          ],
          "title": "H",
        },
        Object {
          "data": Array [
            "III",
          ],
          "title": "I",
        },
        Object {
          "data": Array [
            "JJJ",
          ],
          "title": "J",
        },
        Object {
          "data": Array [
            "KKK",
          ],
          "title": "K",
        },
        Object {
          "data": Array [
            "LLL",
          ],
          "title": "L",
        },
        Object {
          "data": Array [
            "MMM",
          ],
          "title": "M",
        },
        Object {
          "data": Array [
            "NNN",
          ],
          "title": "N",
        },
        Object {
          "data": Array [
            "OOO",
          ],
          "title": "O",
        },
        Object {
          "data": Array [
            "PPP",
          ],
          "title": "P",
        },
        Object {
          "data": Array [
            "QQQ",
          ],
          "title": "Q",
        },
        Object {
          "data": Array [
            "RRR",
          ],
          "title": "R",
        },
        Object {
          "data": Array [
            "SSS",
          ],
          "title": "S",
        },
        Object {
          "data": Array [
            "TTT",
          ],
          "title": "T",
        },
        Object {
          "data": Array [
            "UUU",
          ],
          "title": "U",
        },
        Object {
          "data": Array [
            "VVV",
          ],
          "title": "V",
        },
        Object {
          "data": Array [
            "WWW",
          ],
          "title": "W",
        },
        Object {
          "data": Array [
            "XXX",
          ],
          "title": "X",
        },
        Object {
          "data": Array [
            "YYY",
          ],
          "title": "Y",
        },
        Object {
          "data": Array [
            "ZZZ",
          ],
          "title": "Z",
        },
        Object {
          "data": Array [],
          "title": "#",
        },
      ]
    }
    stickySectionHeadersEnabled={true}
    style={
      Object {
        "flex": 1,
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
  <View
    onLayout={[Function]}
    onMoveShouldSetResponder={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "paddingBottom": 16,
        "paddingTop": 16,
      }
    }
  >
    <View
      key="0"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        A
      </Text>
    </View>
    <View
      key="1"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        B
      </Text>
    </View>
    <View
      key="2"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        C
      </Text>
    </View>
    <View
      key="3"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        D
      </Text>
    </View>
    <View
      key="4"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        E
      </Text>
    </View>
    <View
      key="5"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        F
      </Text>
    </View>
    <View
      key="6"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        G
      </Text>
    </View>
    <View
      key="7"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        H
      </Text>
    </View>
    <View
      key="8"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        I
      </Text>
    </View>
    <View
      key="9"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        J
      </Text>
    </View>
    <View
      key="10"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        K
      </Text>
    </View>
    <View
      key="11"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        L
      </Text>
    </View>
    <View
      key="12"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        M
      </Text>
    </View>
    <View
      key="13"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        N
      </Text>
    </View>
    <View
      key="14"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        O
      </Text>
    </View>
    <View
      key="15"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        P
      </Text>
    </View>
    <View
      key="16"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        Q
      </Text>
    </View>
    <View
      key="17"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        R
      </Text>
    </View>
    <View
      key="18"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        S
      </Text>
    </View>
    <View
      key="19"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        T
      </Text>
    </View>
    <View
      key="20"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        U
      </Text>
    </View>
    <View
      key="21"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        V
      </Text>
    </View>
    <View
      key="22"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        W
      </Text>
    </View>
    <View
      key="23"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        X
      </Text>
    </View>
    <View
      key="24"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        Y
      </Text>
    </View>
    <View
      key="25"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        Z
      </Text>
    </View>
    <View
      key="26"
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        #
      </Text>
    </View>
  </View>
</View>
`;
